rust/src/jsonbuilder.rs

```diff
@@ -98,33 +107,55 @@ pub struct JsonBuilder {
 
 impl JsonBuilder {
     /// Returns a new JsonBuilder in object state.
-    pub fn new_object() -> Self {
-        Self::new_object_with_capacity(INIT_SIZE)
+    pub fn try_new_object() -> Result<Self, JsonError> {
+        Self::try_new_object_with_capacity(INIT_SIZE)
     }
 
-    pub fn new_object_with_capacity(capacity: usize) -> Self {
-        let mut buf = String::with_capacity(capacity);
+    pub fn try_new_object_with_capacity(capacity: usize) -> Result<Self, JsonError> {
+        let mut buf = String::new();
+        buf.try_reserve(capacity)?;
         buf.push('{');
-        Self {
+        Ok(Self {
             buf,
             state: vec![State::None, State::ObjectFirst],
             init_type: Type::Object,
-        }
+        })
     }
 
     /// Returns a new JsonBuilder in array state.
-    pub fn new_array() -> Self {
-        Self::new_array_with_capacity(INIT_SIZE)
+    pub fn try_new_array() -> Result<Self, JsonError> {
+        Self::try_new_array_with_capacity(INIT_SIZE)
     }
 
-    pub fn new_array_with_capacity(capacity: usize) -> Self {
-        let mut buf = String::with_capacity(capacity);
+    pub fn try_new_array_with_capacity(capacity: usize) -> Result<Self, JsonError> {
+        let mut buf = String::new();
+        buf.try_reserve(capacity)?;
         buf.push('[');
-        Self {
+        Ok(Self {
             buf,
             state: vec![State::None, State::ArrayFirst],
             init_type: Type::Array,
+        })
+    }
+
+    /// A wrapper around String::push that pre-allocates data return
+    /// an error if unable to.
+    pub fn push(&mut self, ch: char) -> Result<&mut Self, JsonError> {
+        if self.buf.capacity() == self.buf.len() {
+            self.buf.try_reserve(INIT_SIZE)?;
+        }
+        self.buf.push(ch);
+        Ok(self)
+    }
+
+    /// A wrapper around String::push_str that pre-allocates data
+    /// return an error if unable to.
+    pub fn push_str(&mut self, s: &str) -> Result<&mut Self, JsonError> {
+        if self.buf.capacity() < self.buf.len() + s.len() {
+            self.buf.try_reserve(INIT_SIZE)?;
```

Comment:
I don't like the doubling that many allocators do, but this should at least check try_reserve the max of INIT_SIZE or s.len()...

Reply:
There is no `try_push`, right ?

Reply:
> There is no `try_push`, right ?

No.  But `push`, `push_str` won't allocated unless they need to, which they won't if we `try_reserve` first.